core/profile.c

```diff
@@ -1567,14 +1571,36 @@ void compare_samples(const struct dive *d, const struct plot_info *pi, int idx1,
 			min_depth = data->depth;
 		if (data->depth > max_depth)
 			max_depth = data->depth;
-		/* Try to detect gas changes - this hack might work for some side mount scenarios? */
-		next_pressure = get_plot_pressure(pi, i, 0);
-		if (next_pressure < last_pressure + 2000)
-			bar_used += last_pressure - next_pressure;
+
+		for (unsigned cylinder_index = 0; cylinder_index < pi->nr_cylinders; cylinder_index++) {
+			int next_pressure = get_plot_pressure(pi, i, cylinder_index);
+			if (next_pressure && !start_pressures[cylinder_index])
+				start_pressures[cylinder_index] = next_pressure;
+
+			if (start_pressures[cylinder_index]) {
+				if (last_pressures[cylinder_index]) {
+					bar_used[cylinder_index] += last_pressures[cylinder_index] - next_pressure;
+
+					cylinder_t *cyl = get_cylinder(d, cylinder_index);
```

Comment:
For what I could find it is the same:

https://github.com/mikeller/subsurface/blob/d045c3235c237dbdc782bdbe010f9c170df557dc/core/profile.c#L410

I used `pi->nr_cylinders` because this is what is used in other places in `compare_samples()` - so it seems to be safe to use.

All in all it would probably make sense to eliminate `pi->nr_cylinders` entirely and replace all uses with `d->cylinders.nr` - the duplication seems to only create confusion.

Happy to do this in a separate pull request.

Reply:
@bstoeger:

 > All in all it would probably make sense to eliminate `pi->nr_cylinders` entirely and replace all uses with `d->cylinders.nr` - the duplication seems to only create confusion.
> 
> Happy to do this in a separate pull request.

I had a look into this, and unfortunately this will only _almost_ work in a clean way. The one case that will create awkward code if this is done is access to the contents of the `pi->pressures[]` array:

https://github.com/subsurface/subsurface/blob/a127c4ac63869b753f2912cc3903b0ef68f96ba8/core/profile.h#L121

To me this looks like the best way to eliminate `pi->nr_cylinders` will be to move this to C++ and replace `pi->pressures[]` with a multi-dimensional `std::vector` that can keep track of its size intrinsically.